### PR TITLE
Implement `flake8-future-annotations` FA100

### LINKED ISSUE #3072: Adding flake8-future-annotations
Hi, I authored the flake8-future-annotations plugin: https://github.com/TylerYep/flake8-future-annotations and I think it would be a good addition to ruff.

The goal of the lint rule is to verifies python 3.7+ files use from __future__ import annotations if a type is used in the module that can be rewritten using [PEP 563](https://www.python.org/dev/peps/pep-0563/) e.g. `typing.List, Dict, Optional, Union, etc`.

It pairs especially well with [pyupgrade](https://github.com/asottile/pyupgrade) with the --py37-plus flag or higher, since pyupgrade only replaces type annotations with the PEP 563 rules if from __future__ import annotations is present, which I also see is included in ruff.

I plan to take a stab at implementing it (I see many of the necessary components are already tracked in ast.rs) in the following order:
- Add FA100 (missing future annotations import in Python 3.7+)
- Add an autofix for FA100 (adds missing future annotations import)
- Add FA102 (uses simplified types like `list` or `dict` but missing future annotations import in Python 3.7 to 3.10)

All feedback is welcome, thanks!

### PR DESCRIPTION
**Pull request by TylerYep**
This PR tackles the first part of #3072, which is to raise a lint error when a rewrite-able type is used without `from __future__ import annotations` present.

Test cases taken from: https://github.com/TylerYep/flake8-future-annotations

Some of the test cases I added (prefixed by `no_future`) do not raise lint warnings now (as expected), but will raise lint warnings once FA102 is implemented. I am adding them to ensure they do not raise extraneous errors to begin with.

This lint rule also found some available fixes in `scripts/check_ecosystem.py`, which tells me there's some value to having this around :)

The next two PRs will add the autofix functionality for FA100 and FA101 and/or FA102. Thanks for the review!

### CODE REVIEW

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff_python_stdlib/src/typing.rs
Comment (charliermarsh) [inline crates/ruff_python_stdlib/src/typing.rs:204]: @TylerYep - Could you speak a bit towards how this list was determined? I initially thought it'd only include builtin collections (like `List` and `Dict`), but including `DefaultDict` and `Deque` means we're asking users to import those from `collections`. So should it also include `AsyncIterator`, `Mapping`, etc. -- all the things that were moved from `typing` to `collections.abc` in Python 3.9? (I think `crates/ruff/src/rules/pyupgrade/rules/deprecated_import.rs` is our most comprehensive accounting of those moves, for reference.)

Comment (TylerYep) [reply crates/ruff_python_stdlib/src/typing.rs:204]: These are all of the data structures that can be rewritten by pyupgrade into lowercase form e.g.

List[int] becomes list[int]

DefaultDict[str, bool] becomes defaultdict[str, bool]

Comment (charliermarsh) [reply crates/ruff_python_stdlib/src/typing.rs:204]: By that logic, should `typing.AsyncIterator` and friends also be included?

Comment (charliermarsh) [reply crates/ruff_python_stdlib/src/typing.rs:204]: To my knowledge, `pyupgrade` only rewrites the builtins: https://github.com/asottile/pyupgrade/blob/7c551d39dfa69851607d85392ffa296568e92436/pyupgrade/_plugins/typing_pep585.py#L15

Comment (TylerYep) [reply crates/ruff_python_stdlib/src/typing.rs:204]: Ah, in that case we don't need to include Deque, DefaultDict, or FrozenSet. Fixing...

**Comment by TylerYep** [comment]
I'm not sure why tests are failing - running `cargo test` locally passes cleanly for me.

Any idea @charliermarsh ?

**Comment by charliermarsh** [comment]
@TylerYep - Just needed to merge, we recently hit a limit on a fixed-size vector that stores all the rules and it had to be increased.

**Review by MichaReiser — APPROVED** [review]

**Comment by TylerYep** [comment]
I went back to my original plugin and remembered why DefaultDict and Deque were originally included in FA100:

The main goal of this plugin is to signal to users when a type annotation could be rewritten in a more concise way when `from __future__ import annotations` is added.

1. This code could be simplified (and is currently autofixed by Pyupgrade):
```
from typing import List, Optional, Union

a: List[int] = []
b: Optional[Union[int, str]] = None
```

```
from __future__ import annotations

a: list[int] = []
b: int | str | None = None
```

2. This code could be simplified (but is not currently autofixed by Pyupgrade):

```
from collections import defaultdict, deque
from typing import DefaultDict, Deque

a: DefaultDict[int] = defaultdict(int)
b: Deque[int] = deque()
```

```
from __future__ import annotations

from collections import defaultdict, deque

a: defaultdict[int] = defaultdict(int)
b: deque[int] = deque()
```


https://github.com/asottile/pyupgrade/issues/799 provides context on why collection types are not rewritten using Pyupgrade (import aliasing is hard). It might be possible to extend the autofix functionality to rewrite all of them correctly, but that is out of scope for now.

Afaik, DefaultDict and Deque are the only two code changes included in PEP585, improved using `from __future__ import annotations`, but not autofixed by Pyupgrade.

3. This code cannot be simplified, but can use a different import. These cases are not intended to be raised by this flake8 rule because the type annotation doesn't need to be changed by adding `from __future__ import annotations`:
```
from typing import AsyncIterator, Mapping

a: Mapping[int, str] = {}
b: AsyncIterator = ...
```

```
from collections.abc import AsyncIterator, Mapping

a: Mapping[int, str] = {}
b: AsyncIterator = ...

```

My view is that both 1 and 2 should be raised by this plugin, but not 3, because the code itself is not improved by adding `from __future__ import annotations`.

If this makes more sense to you, we can revert 4373751.

File: crates/ruff/src/rules/flake8_future_annotations/rules.rs
Comment (calumy) [inline crates/ruff/src/rules/flake8_future_annotations/rules.rs]: This section should not have been removed when updating the docs. I'll update the error message to say something along the lines of "Rule `missing-future-annotations-with-imports` docs are not formatted. The example section should be rewritten to:"

Comment (calumy) [reply crates/ruff/src/rules/flake8_future_annotations/rules.rs]: See #4312

Comment (TylerYep) [reply crates/ruff/src/rules/flake8_future_annotations/rules.rs]: Thanks, fixed! The error message looked copy+pastable but did not contain the full output. Hopefully that can be an autofix via pre-commit

File: crates/ruff/src/rules/flake8_future_annotations/rules.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_future_annotations/rules.rs:94]: Sorry, last question: what's the benefit in flagging both the usages and the imports? Would flagging the import alone be insufficient?

Comment (TylerYep) [reply crates/ruff/src/rules/flake8_future_annotations/rules.rs:94]: If the user uses `import typing`, we can only flag `typing.List` at the usage. If they use `from typing import List`, then flagging the import alone is sufficient.

**Comment by charliermarsh** [comment]
Will merge this after cutting the release.

File: crates/ruff/src/checkers/ast/mod.rs
Comment (charliermarsh) [inline crates/ruff/src/checkers/ast/mod.rs:2451]: These should only be activated for minimum-version < Python 3.9, right? Since PEP 585 was part of Python 3.9 (so future annotations aren't required to use the standard-library variants in 3.9 and later).

Comment (TylerYep) [reply crates/ruff/src/checkers/ast/mod.rs:2460]: In practice, this works on Python 3.7+.

Comment (charliermarsh) [reply crates/ruff/src/checkers/ast/mod.rs:2460]: But if your minimum-supported Python version is Python 3.9, you don't need `__future__` annotations to use the standard-library generics. So these errors would already be detected and fixed by the existing `pyupgrade` rules, right?

Comment (TylerYep) [reply crates/ruff/src/checkers/ast/mod.rs:2460]: Ah, you're asking about the upper version. Unions like `str | None` didn't start working until Python 3.10+, which are also covered by this plugin. After 3.10 the future annotations import is not as useful anymore.

**Comment by calumy** [comment]
> Will merge this after cutting the release.

I thought it might be worth flagging that the changes to the ecosystem check seem to be causing it to error. I don't have time to investigate now but thought it would be worth saying something before this is merged.

**Comment by TylerYep** [comment]
What is the error?

**Comment by calumy** [comment]
> What is the error? I see check_ecosystem imports Self, which must mean it's running on Python 3.11+, so none of my changes should impact it.


Error in the ecosystem check above:
```
'async_generator' object does not support the asynchronous context manager protocol

```

**Comment by TylerYep** [comment]
Was that caused by removing asynccontextmanager? It was an unused import that pre-commit told me to remove

File: scripts/check_ecosystem.py
Comment (charliermarsh) [inline scripts/check_ecosystem.py:39]: It looks like the usage was removed here?

Comment (TylerYep) [reply scripts/check_ecosystem.py:39]: Thanks, fixed the bad rebase.

**Comment by charliermarsh** [comment]
I made a few changes here that I _think_ are improvements based on some of the capabilities we have in Ruff, but if you disagree, just let me know and we can of course discuss further -- want to merge to keep momentum.

1. I removed the diagnostics that trigger on-import, and instead, I'm more closely following our pyupgrade rules. We flag all usages, regardless of whether the user does `import typing` or `from typing import List` -- when the user writes `typing.List` or `List`, we resolve it to `typing.List`. So it's always sufficient to just flag usages, given that we have enough binding tracking and semantic analysis to resolve those usages, regardless of how they're imported.
2. These rules now only flag if your minimum version is below Python 3.9 (or Python 3.10, for `Optional` and `Union`), since adding `__future__` on those versions has no effect with regards to simplifying the imports. (You can already use the standard library generics on those versions.)
3. We weren't flagging `Optional` or `Union`, so I added that too.

File: crates/ruff/src/checkers/ast/mod.rs
Comment (charliermarsh) [inline crates/ruff/src/checkers/ast/mod.rs:2344]: This condition is a lot to read, I'd like to clean this up... But the idea is: mirror the logic below for `Rule::NonPEP585Annotation`, but flip the `!self.ctx.future_annotations()`. So we're asking: would we trigger here if `__future__` annotations were enabled?

**Pull request merged**
